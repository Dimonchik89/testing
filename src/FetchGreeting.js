// import { useState } from 'react'
// import axios from 'axios'

// // пропом компонента является адрес конечной точки
// // для получения приветствия от сервера
// const FetchGreeting = ({ url }) => {
//   // состояние приветствия
//   const [greeting, setGreeting] = useState('')
//   // состояние ошибки
//   const [error, setError] = useState(null)
//   // состояние нажатия кнопки
//   const [btnClicked, setBtnClicked] = useState(false)

//   // метод для получения приветствия от сервера
//   const fetchGreeting = (url) =>
//     axios
//       .get(url)
//       // если запрос выполнен успешно
//       .then((res) => {
//         const { data } = res
//         const { greeting } = data
//         setGreeting(greeting)
//         setBtnClicked(true)
//       })
//       // если возникла ошибка
//       .catch((e) => {
//         setError(e)
//       })

//   // текст кнопки
//   const btnText = btnClicked ? 'Готово' : 'Получить приветствие'

//   return (
//     <div>
//       <button onClick={() => fetchGreeting(url)} disabled={btnClicked}>
//         {btnText}
//       </button>
//       {/* если запрос выполнен успешно */}
//       {greeting && <h1>{greeting}</h1>}
//       {/* если возникла ошибка */}
//       {error && <p role='alert'>Не удалось получить приветствие</p>}
//     </div>
//   )
// }

// export default FetchGreeting


import { useState } from 'react'
import axios from 'axios'
import { useGreetingContext } from './GreetingProvider'

const FetchGreeting = ({ url }) => {
  // извлекаем состояние и операции из контекста
  const { state, actions } = useGreetingContext()
  const [btnClicked, setBtnClicked] = useState(false)

  const fetchGreeting = (url) =>
    axios
      .get(url)
      .then((res) => {
        const { data } = res
        const { greeting } = data
        // !
        actions.setSuccess(greeting)
        setBtnClicked(true)
      })
      .catch((e) => {
        // !
        actions.setError(e)
      })

  const btnText = btnClicked ? 'Готово' : 'Получить приветствие'

  return (
    <div>
      <button onClick={() => fetchGreeting(url)} disabled={btnClicked}>
        {btnText}
      </button>
      {/* ! */}
      {state.greeting && <h1 data-cy='heading'>{state.greeting}</h1>}
      {state.error && <p role='alert'>Не удалось получить приветствие</p>}
    </div>
  )
}

export default FetchGreeting