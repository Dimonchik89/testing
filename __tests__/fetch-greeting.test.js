import { rest   } from "msw";
import { setupServer } from 'msw/node'
// import { render, fireEvent, waitFor, screen } from '@testing-library/react'
import { render, fireEvent, waitFor, screen } from 'test-utils'
import FetchGreeting from '../src/FetchGreeting'
import userEvent from '@testing-library/user-event'

import "@testing-library/jest-dom";

const server = setupServer(
  rest.get('/greeting', (req, res, ctx) =>
    res(ctx.json({ greeting: 'Привет!' }))
  )
)

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

describe("Home", () => {
    test("успешное получение приветствия", async () => {
        // render(<FetchGreeting url='/greeting'/>)
        
        const { container, getByDataCy } = render(<FetchGreeting url='/greeting' />)
        // тестируем текущее состояние `UI` с помощью снимка
        expect(container).toMatchSnapshot()

        fireEvent.click(screen.getByText('Получить приветствие'))
        await waitFor(() => screen.getByRole("heading"))
        // состояние `UI` изменилось, поэтому нужен еще один снимок
        expect(container).toMatchSnapshot()

        expect(getByDataCy('heading')).toHaveTextContent('Привет!')
        expect(screen.getByRole("button")).toHaveTextContent("Готово")
        expect(screen.getByRole("button")).toBeDisabled()
        // expect(screen.getByRole("heading")).toHaveTextContent('Привет!')
        // expect(screen.getByRole("button")).toHaveTextContent("Готово")
        // expect(screen.getByRole("button")).toBeDisabled()
    })

    test("обработка ошибки сервера", async () => {
        server.use(rest.get('/greeting', (req, res, cts) => res(cts.status(500))))
        // render(<FetchGreeting url="greeting"/>)
        const { container } = render(<FetchGreeting url="/greeting"/>)

        expect(container).toMatchSnapshot()

        // const user = userEvent.setup()
        await userEvent.click(screen.getByRole("button"))
        await waitFor(() => screen.getByRole("alert"))

        expect(container).toMatchSnapshot()

        expect(screen.getByRole("alert")).toHaveTextContent('Не удалось получить приветствие')
        expect(screen.getByRole("button")).not.toBeDisabled()
    })
})

test.todo('получение приветствия')