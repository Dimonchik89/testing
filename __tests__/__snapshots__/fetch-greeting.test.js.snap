// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home обработка ошибки сервера 1`] = `
<div>
  <div>
    <button
      data-cy="button"
    >
      Получить приветствие
    </button>
  </div>
</div>
`;

exports[`Home обработка ошибки сервера 2`] = `
<div>
  <div>
    <button
      data-cy="button"
    >
      Получить приветствие
    </button>
    <p
      role="alert"
    >
      Не удалось получить приветствие
    </p>
  </div>
</div>
`;

exports[`Home успешное получение приветствия 1`] = `
<div>
  <div>
    <button
      data-cy="button"
    >
      Получить приветствие
    </button>
  </div>
</div>
`;

exports[`Home успешное получение приветствия 2`] = `
<div>
  <div>
    <button
      data-cy="button"
      disabled=""
    >
      Готово
    </button>
    <h1
      data-cy="heading"
    >
      Привет!
    </h1>
  </div>
</div>
`;
